src/Unit-3/lesson4/CollaborativeCategoryEditing/EditCategoryForm.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Reactive.Linq;
+using System.Windows.Forms;
+
+using Raven.Abstractions.Data;
+using Raven.Client;
+
+namespace CollaborativeCategoryEditing
+{
+    public partial class EditCategoryForm : Form
+    {
+        public EditCategoryForm()
+        {
+            InitializeComponent();
+        }
+
+        private IDocumentSession _session;
+        private Category _category;
+        private IDisposable _subscription;
+        private Etag _localEtag;
+
+        private void LoadAndSubscribeButton_Click(object sender, EventArgs e)
+        {
+            _session = DocumentStoreHolder.Store.OpenSession();
+            _category = _session.Load<Category>(CategoryIdTextbox.Text);
+
+            if (_category == null)
+            {
+                MessageBox.Show("Category not found!");
+                _session.Dispose();
+            }
+            else
+            {
+                NameTextbox.Text = _category.Name;
+                DescriptionTextbox.Text = _category.Description;
+
+                _localEtag = _session.Advanced.GetEtagFor(_category);
+                _subscription = DocumentStoreHolder.Store
+                    .Changes()
+                    .ForDocument(CategoryIdTextbox.Text)
+                    .Where(DocumentChangedByOtherUser)
+                    .Subscribe(DocumentChangedOnServer);
+
+                ToggleEditing();
+            }
+        }
+
+        private bool DocumentChangedByOtherUser(DocumentChangeNotification change)
+        {
+            if (_savesCount == 0) return true;
+
+            var numberOfServerChanges = change.Etag.Changes - _localEtag.Changes;
```

Comment:
Once we loose the connection we try to renew it then so the client will be notified even after the restart.